### ISSUE TITLE
false positive E037

### ISSUE THREAD

**Issue by kdy1** [rust-lang/rust#68636]
While investigation #68564, I found a real bug.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c959b9e4ee176c4430586425307787ec

```rust
pub trait Fold<T: ?Sized> {
    fn fold(&mut self, node: T) -> T;
}

struct A;
struct B;

impl Fold<A> for dyn Fold<B> {}
```
It results in an error even on stable.

`Fold<A>` and `Fold<B>` is treated as a separate trait (#68564) while resolving, but treated as a same trait while verifying impl blocks.

```
   Compiling playground v0.0.1 (/playground)
error[E0371]: the object type `(dyn Fold<B> + 'static)` automatically implements the trait `Fold`
  --> src/lib.rs:11:1
   |
11 | impl Fold<A> for dyn Fold<B> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `(dyn Fold<B> + 'static)` automatically implements trait `Fold`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0371`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.
```

**Comment by kdy1** [comment]
While chatting on rust discord, someone suggested that it conflicts if a type implements `Fold<A>` and `Fold<B>`. I'm opening #68564 again and I'll explain more on the issue.

**Comment by sffc** [comment]
I'm having this issue.  My workaround is to make an intermediate trait, like this:

```rust
// Same functions as Fold<B>
trait FoldB {
    fn fold(&mut self, node: B) -> B;
}

// Auto-implement FoldB for Fold<B>
impl<T> FoldB for T where T: Fold<B> {
    fn fold(&mut self, node: B) -> B {
        Fold::<B>::fold(self, node)
    }
}

impl Fold<A> for dyn FoldB {
    // works!
}
```

Updated playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3ad915417abd9945d362ee6f04475a79
